src/Domain/CoordinateMaps/CylindricalEndcap.hpp

```diff
@@ -0,0 +1,150 @@
+ * possible to construct a valid map without this assumption, but the
+ * assumption simplifies the code, and the expected use cases obey
+ * this restriction.
+ *
+ * We also demand that \f$z_\mathrm{P} > C_1^2\f$, that is, the plane
+ * in the above and below figures lies to the right of \f$C_1^2\f$.
+ * This restriction not strictly necessary but is made for simplicity.
+ *
+ * The map is invertible only for some choices of the projection point
+ * \f$P\f$.  Given the above restrictions, the allowed values of
+ * \f$P\f$ are illustrated by the following diagram:
+ *
+ * \image html CylindricalEndcap_Allowed.svg "Allowed region for P." width=75%
+ *
+ * The plane \f$z=z_\mathrm{P}\f$ intersects sphere 1 on a circle. The
+ * cone with apex \f$C_1\f$ that intersects that circle has opening
+ * angle \f$2\theta\f$ as shown in the above figure. Construct another
+ * cone, the "invertibility cone", with apex \f$S\f$ chosen such that
+ * the two cones intersect at right angles on the circle; thus the
+ * opening angle of the invertibility cone is \f$2(\pi-\theta)\f$.  A
+ * necessary condition for invertibility is that the projection point
+ * \f$P\f$ lies inside the invertibility cone, but not between \f$S\f$
+ * and sphere 1.  Placing the projection point \f$P\f$ to the right of
+ * \f$S\f$ (but inside the invertibility cone) is ok for
+ * invertibility.
+ *
+ * In addition to invertibility and the two additional restrictions
+ * already mentioned above, we demand a few more restrictions on the
+ * map parameters to simplify the logic for the expected use cases, and
+ * to ensure that jacobians do not get too large. We demand:
+ *
+ * - \f$P\f$ is not too close to the edge of the invertibility cone.
+ * - \f$P\f$ is contained in sphere 2.
+ * - \f$P\f$ is to the left of \f$z_\mathrm{P}\f$.
+ * - \f$z_\mathrm{P}\f$ is not too close to the center or the edge of sphere 1.
+ * - If a line segment is drawn between \f$P\f$ and any point on the
+ *   intersection circle, the angle between the line segment and
+ *   the x-axis is smaller than \f$\pi/3\f$.
+ *
+ */
+class CylindricalEndcap {
+ public:
+  static constexpr size_t dim = 3;
+  CylindricalEndcap(const std::array<double, 3>& center_one,
+                       const std::array<double, 3>& center_two,
+                       const std::array<double, 3>& proj_center,
+                       double radius_one, double radius_two,
+                       double z_plane) noexcept;
+
+  CylindricalEndcap() = default;
+  ~CylindricalEndcap() = default;
+  CylindricalEndcap(CylindricalEndcap&&) = default;
+  CylindricalEndcap(const CylindricalEndcap&) = default;
+  CylindricalEndcap& operator=(const CylindricalEndcap&) = default;
+  CylindricalEndcap& operator=(CylindricalEndcap&&) = default;
+
+  template <typename T>
+  std::array<tt::remove_cvref_wrap_t<T>, 3> operator()(
+      const std::array<T, 3>& source_coords) const noexcept;
+
+  boost::optional<std::array<double, 3>> inverse(
```

Comment:
why is this one not templated on type T like operator(), Jacobian, invjacobian?

Reply:
Because the inverse is always done pointwise in all the maps, and in the base class. If it were done with the entire array, we would need a better way than boost::optional to signal whether the inverse succeeds, since it can succeed for some points and not others.